---
page: jamboree
navText: HTAN Data Jamboree
---

<meta content="text/html; charset=UTF-8" http-equiv="content-type" /><style
    type="text/css"
>
    .lst-kix_list_5-1 > li {
        counter-increment: lst-ctn-kix_list_5-1;
    }
    .lst-kix_list_1-1 > li {
        counter-increment: lst-ctn-kix_list_1-1;
    }
    .lst-kix_list_3-0 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_3-1 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_3-2 > li:before {
        content: '\0025aa   ';
    }
    ol.lst-kix_list_1-8.start {
        counter-reset: lst-ctn-kix_list_1-8 0;
    }
    .lst-kix_list_5-0 > li {
        counter-increment: lst-ctn-kix_list_5-0;
    }
    .lst-kix_list_3-5 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_3-4 > li:before {
        content: 'o  ';
    }
    ol.lst-kix_list_1-5.start {
        counter-reset: lst-ctn-kix_list_1-5 0;
    }
    .lst-kix_list_3-3 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_3-8 > li:before {
        content: '\0025aa   ';
    }
    ol.lst-kix_list_5-3.start {
        counter-reset: lst-ctn-kix_list_5-3 0;
    }
    .lst-kix_list_3-6 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_3-7 > li:before {
        content: 'o  ';
    }
    ol.lst-kix_list_5-0.start {
        counter-reset: lst-ctn-kix_list_5-0 0;
    }
    .lst-kix_list_1-2 > li {
        counter-increment: lst-ctn-kix_list_1-2;
    }
    .lst-kix_list_5-2 > li {
        counter-increment: lst-ctn-kix_list_5-2;
    }
    .lst-kix_list_5-0 > li:before {
        content: '' counter(lst-ctn-kix_list_5-0, decimal) '. ';
    }
    .lst-kix_list_5-4 > li {
        counter-increment: lst-ctn-kix_list_5-4;
    }
    .lst-kix_list_1-4 > li {
        counter-increment: lst-ctn-kix_list_1-4;
    }
    ol.lst-kix_list_1-6.start {
        counter-reset: lst-ctn-kix_list_1-6 0;
    }
    .lst-kix_list_4-8 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_5-3 > li:before {
        content: '' counter(lst-ctn-kix_list_5-3, decimal) '. ';
    }
    .lst-kix_list_4-7 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_5-2 > li:before {
        content: '' counter(lst-ctn-kix_list_5-2, lower-roman) '. ';
    }
    .lst-kix_list_5-1 > li:before {
        content: '' counter(lst-ctn-kix_list_5-1, lower-latin) '. ';
    }
    ul.lst-kix_list_4-8 {
        list-style-type: none;
    }
    .lst-kix_list_5-7 > li:before {
        content: '' counter(lst-ctn-kix_list_5-7, lower-latin) '. ';
    }
    ol.lst-kix_list_5-6.start {
        counter-reset: lst-ctn-kix_list_5-6 0;
    }
    ul.lst-kix_list_4-6 {
        list-style-type: none;
    }
    .lst-kix_list_5-6 > li:before {
        content: '' counter(lst-ctn-kix_list_5-6, decimal) '. ';
    }
    .lst-kix_list_5-8 > li:before {
        content: '' counter(lst-ctn-kix_list_5-8, lower-roman) '. ';
    }
    ul.lst-kix_list_4-7 {
        list-style-type: none;
    }
    ul.lst-kix_list_4-0 {
        list-style-type: none;
    }
    ul.lst-kix_list_4-1 {
        list-style-type: none;
    }
    .lst-kix_list_5-4 > li:before {
        content: '' counter(lst-ctn-kix_list_5-4, lower-latin) '. ';
    }
    ul.lst-kix_list_4-4 {
        list-style-type: none;
    }
    .lst-kix_list_5-5 > li:before {
        content: '' counter(lst-ctn-kix_list_5-5, lower-roman) '. ';
    }
    ul.lst-kix_list_4-5 {
        list-style-type: none;
    }
    ul.lst-kix_list_4-2 {
        list-style-type: none;
    }
    ul.lst-kix_list_4-3 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-0.start {
        counter-reset: lst-ctn-kix_list_1-0 0;
    }
    ol.lst-kix_list_5-5.start {
        counter-reset: lst-ctn-kix_list_5-5 0;
    }
    ol.lst-kix_list_1-3 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-0 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-4 {
        list-style-type: none;
    }
    .lst-kix_list_2-6 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_2-7 > li:before {
        content: 'o  ';
    }
    ol.lst-kix_list_5-1 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-5 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-2 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-6 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-0 {
        list-style-type: none;
    }
    .lst-kix_list_2-4 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_2-5 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_2-8 > li:before {
        content: '\0025aa   ';
    }
    ol.lst-kix_list_1-1 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-2 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-4.start {
        counter-reset: lst-ctn-kix_list_5-4 0;
    }
    ul.lst-kix_list_3-7 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-8 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-1.start {
        counter-reset: lst-ctn-kix_list_5-1 0;
    }
    ol.lst-kix_list_5-7 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-1 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-8 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-2 {
        list-style-type: none;
    }
    .lst-kix_list_5-7 > li {
        counter-increment: lst-ctn-kix_list_5-7;
    }
    ul.lst-kix_list_3-0 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-3 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-7 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-5 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-4 {
        list-style-type: none;
    }
    .lst-kix_list_1-7 > li {
        counter-increment: lst-ctn-kix_list_1-7;
    }
    ol.lst-kix_list_1-8 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-6 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-5 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-3 {
        list-style-type: none;
    }
    ol.lst-kix_list_5-6 {
        list-style-type: none;
    }
    ul.lst-kix_list_3-4 {
        list-style-type: none;
    }
    .lst-kix_list_5-8 > li {
        counter-increment: lst-ctn-kix_list_5-8;
    }
    .lst-kix_list_4-0 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_4-1 > li:before {
        content: 'o  ';
    }
    ol.lst-kix_list_1-7.start {
        counter-reset: lst-ctn-kix_list_1-7 0;
    }
    .lst-kix_list_4-4 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_1-5 > li {
        counter-increment: lst-ctn-kix_list_1-5;
    }
    .lst-kix_list_4-3 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_4-5 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_4-2 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_4-6 > li:before {
        content: '\0025cf   ';
    }
    ol.lst-kix_list_5-7.start {
        counter-reset: lst-ctn-kix_list_5-7 0;
    }
    .lst-kix_list_1-8 > li {
        counter-increment: lst-ctn-kix_list_1-8;
    }
    ol.lst-kix_list_1-4.start {
        counter-reset: lst-ctn-kix_list_1-4 0;
    }
    .lst-kix_list_5-5 > li {
        counter-increment: lst-ctn-kix_list_5-5;
    }
    ol.lst-kix_list_1-1.start {
        counter-reset: lst-ctn-kix_list_1-1 0;
    }
    .lst-kix_list_5-3 > li {
        counter-increment: lst-ctn-kix_list_5-3;
    }
    ol.lst-kix_list_1-3.start {
        counter-reset: lst-ctn-kix_list_1-3 0;
    }
    ul.lst-kix_list_2-8 {
        list-style-type: none;
    }
    ol.lst-kix_list_1-2.start {
        counter-reset: lst-ctn-kix_list_1-2 0;
    }
    ul.lst-kix_list_2-2 {
        list-style-type: none;
    }
    .lst-kix_list_1-0 > li:before {
        content: '' counter(lst-ctn-kix_list_1-0, decimal) '. ';
    }
    ul.lst-kix_list_2-3 {
        list-style-type: none;
    }
    ul.lst-kix_list_2-0 {
        list-style-type: none;
    }
    ul.lst-kix_list_2-1 {
        list-style-type: none;
    }
    ul.lst-kix_list_2-6 {
        list-style-type: none;
    }
    .lst-kix_list_1-1 > li:before {
        content: '' counter(lst-ctn-kix_list_1-1, lower-latin) '. ';
    }
    .lst-kix_list_1-2 > li:before {
        content: '' counter(lst-ctn-kix_list_1-2, lower-roman) '. ';
    }
    ul.lst-kix_list_2-7 {
        list-style-type: none;
    }
    ul.lst-kix_list_2-4 {
        list-style-type: none;
    }
    ul.lst-kix_list_2-5 {
        list-style-type: none;
    }
    .lst-kix_list_1-3 > li:before {
        content: '' counter(lst-ctn-kix_list_1-3, decimal) '. ';
    }
    .lst-kix_list_1-4 > li:before {
        content: '' counter(lst-ctn-kix_list_1-4, lower-latin) '. ';
    }
    .lst-kix_list_1-0 > li {
        counter-increment: lst-ctn-kix_list_1-0;
    }
    .lst-kix_list_1-6 > li {
        counter-increment: lst-ctn-kix_list_1-6;
    }
    .lst-kix_list_1-7 > li:before {
        content: '' counter(lst-ctn-kix_list_1-7, lower-latin) '. ';
    }
    ol.lst-kix_list_5-8.start {
        counter-reset: lst-ctn-kix_list_5-8 0;
    }
    .lst-kix_list_1-3 > li {
        counter-increment: lst-ctn-kix_list_1-3;
    }
    .lst-kix_list_1-5 > li:before {
        content: '' counter(lst-ctn-kix_list_1-5, lower-roman) '. ';
    }
    .lst-kix_list_1-6 > li:before {
        content: '' counter(lst-ctn-kix_list_1-6, decimal) '. ';
    }
    li.li-bullet-0:before {
        margin-left: -18pt;
        white-space: nowrap;
        display: inline-block;
        min-width: 18pt;
    }
    .lst-kix_list_5-6 > li {
        counter-increment: lst-ctn-kix_list_5-6;
    }
    .lst-kix_list_2-0 > li:before {
        content: '\0025cf   ';
    }
    .lst-kix_list_2-1 > li:before {
        content: 'o  ';
    }
    .lst-kix_list_1-8 > li:before {
        content: '' counter(lst-ctn-kix_list_1-8, lower-roman) '. ';
    }
    .lst-kix_list_2-2 > li:before {
        content: '\0025aa   ';
    }
    .lst-kix_list_2-3 > li:before {
        content: '\0025cf   ';
    }
    ol.lst-kix_list_5-2.start {
        counter-reset: lst-ctn-kix_list_5-2 0;
    }
    ol {
        margin: 0;
        padding: 0;
    }
    table td,
    table th {
        padding: 0;
    }
    .c11 {
        margin-left: 36pt;
        padding-top: 0pt;
        padding-left: 0pt;
        padding-bottom: 0pt;
        line-height: 1.0791666666666666;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c2 {
        margin-left: 36pt;
        padding-top: 0pt;
        padding-left: 0pt;
        padding-bottom: 6pt;
        line-height: 1.0999999999999999;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c18 {
        margin-left: 36pt;
        padding-top: 0pt;
        padding-bottom: 6pt;
        line-height: 1.0999999999999999;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c13 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-style: normal;
    }
    .c0 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-style: normal;
    }
    .c16 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-style: italic;
    }
    .c6 {
        color: #000000;
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-style: normal;
    }
    .c4 {
        padding-top: 0pt;
        padding-bottom: 8pt;
        line-height: 1.0791666666666666;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c34 {
        padding-top: 0pt;
        padding-bottom: 0pt;
        line-height: 1.0999999999999999;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c3 {
        padding-top: 0pt;
        padding-bottom: 8pt;
        line-height: 1.0791666666666666;
        orphans: 2;
        widows: 2;
    }
    .c5 {
        color: #000000;
        font-weight: 700;
        text-decoration: none;
        vertical-align: baseline;
        font-style: normal;
    }
    .c28 {
        padding-top: 0pt;
        padding-bottom: 6pt;
        line-height: 1;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c19 {
        padding-top: 0pt;
        padding-bottom: 6pt;
        line-height: 1.0999999999999999;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .c20 {
        font-weight: 400;
        text-decoration: none;
        vertical-align: baseline;
        font-style: normal;
    }
    .c24 {
        font-weight: 400;
        vertical-align: baseline;
        font-style: normal;
    }
    .c15 {
        text-decoration-skip-ink: none;
        -webkit-text-decoration-skip: none;
        color: #0563c1;
    }
    .c21 {
        text-decoration-skip-ink: none;
        -webkit-text-decoration-skip: none;
        color: #1155cc;
    }
    .c25 {
        padding-top: 0pt;
        padding-bottom: 0pt;
        line-height: 1;
        text-align: left;
    }
    .c37 {
        background-color: #ffffff;
        max-width: 468pt;
        padding: 72pt 72pt 72pt 72pt;
    }
    .c27 {
        margin-left: 36pt;
        padding-left: 0pt;
    }
    .c7 {
        color: inherit;
        text-decoration: inherit;
    }
    .c14 {
        border: 1px solid black;
        margin: 5px;
    }
    .c36 {
        vertical-align: baseline;
    }
    .c12 {
        padding: 0;
        margin: 0;
    }
    .c33 {
        /*color: #ff0000;*/
        color: #6237e5;
    }
    .c22 {
        font-size: 20pt;
    }
    .c35 {
        font-size: 16pt;
    }
    .c9 {
        color: #444444;
    }
    .c29 {
        text-decoration: none;
    }
    .c17 {
        height: 11pt;
    }
    .c30 {
        color: #000000;
    }
    .c8 {
        font-weight: 700;
    }
    .c10 {
        font-size: 15pt;
    }
    .c31 {
        /*color: #980000;*/
        color: #6237e5;
    }
    .c32 {
        font-size: 11pt;
    }
    .c23 {
        font-style: italic;
    }
    .c26 {
        color: #2f5496;
    }
    .title {
        padding-top: 24pt;
        color: #000000;
        font-weight: 700;
        font-size: 36pt;
        padding-bottom: 6pt;
        line-height: 1.0791666666666666;
        page-break-after: avoid;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    .subtitle {
        padding-top: 18pt;
        color: #666666;
        font-size: 24pt;
        padding-bottom: 4pt;
        line-height: 1.0791666666666666;
        page-break-after: avoid;
        font-style: italic;
        orphans: 2;
        widows: 2;
        text-align: left;
    }
    li {
        color: #000000;
        font-size: 11pt;
    }
    p {
        margin: 0;
        color: #000000;
        font-size: 11pt;
    }
    .image-container {
        background-color: rgba(
            8,
            3,
            8,
            0.5
        ); /* Adjust the alpha value to control the darkness */
        width: 18%;
        filter: brightness(100%) sepia(90%) hue-rotate(270deg) saturate(500%)
            contrast(0.8);
        float: right; /* Float the image to the left */
        margin: 5px 0px 5px 15px; /* Add some margin for spacing */
    }
</style>
<body>
    <div>
        <img class="image-container" src="/jamboree.png" />
        <p class="c3">
            <span class="c22 c8">Virtual </span
            ><span class="c8 c22">HTAN Data Jamboree</span>
        </p>
        <p class="c3"><span class="c33 c8 c35">December 4-7, 2023</span></p>
        <p class="c3">
            <span class="c8 c10"
                >Now open for application: submit your
                <a class="c31" href="https://forms.gle/V2XW6bm2KJJbX6CF6"
                    >application</a
                >
                by </span
            ><span class="c8 c10 c33">October 24, 2023</span
            ><span class="c5 c10"></span>
        </p>
    </div>
    <p class="c4 c17"><span class="c0"></span></p>
    <p class="c4">
        <span
            >The National Cancer Institute of the NIH is pleased to announce a
            virtual Human Tumor Atlas Network (HTAN) Data Jamboree focused on
            utilizing </span
        ><span class="c1 c15"
            ><a class="c7" href="https://data.humantumoratlas.org/"
                >HTAN atlas datasets</a
            ></span
        ><span class="c9 c1">&nbsp;</span><span class="c1">within the </span
        ><span class="c15 c1"
            ><a class="c7" href="https://humantumoratlas.org/explore"
                >HTAN Data Portal</a
            ></span
        ><span class="c1 c9">&nbsp;</span><span class="c1">and</span
        ><span class="c1">&nbsp;</span
        ><span class="c15 c1"
            ><a class="c7" href="https://datacommons.cancer.gov/"
                >NCI Cancer Research Data Commons (CRDC)</a
            ></span
        ><span class="c9 c1">&nbsp;</span
        ><span class="c1">resources accessible from the </span
        ><span class="c1 c21"
            ><a class="c7" href="https://www.cancergenomicscloud.org/datasets"
                >Seven Bridges-Cancer Genomics Cloud</a
            ></span
        ><span class="c9 c1">.</span><span class="c24 c9 c1 c29">&nbsp;</span>
    </p>
    <p class="c4"></p>
    <p class="c4">
        <span class="c5">The Opportunity and A Call to Action</span>
    </p>
    <p class="c4">
        <span class="c1"
            >Single-cell and spatial omics technologies are becoming widely used
            across biomedical research including in cancer. These rich and
            complex datasets provide an untapped resource for reuse, pan-cancer
            analysis, and integration with other datasets. However, the best
            approaches and tools for reusing these datasets are not well
            defined. </span
        ><span class="c1 c29 c30 c8 c23 c36"
            >Do you have an interest in utilizing single-cell and spatial omics
            approaches in your cancer research? Do you have creative approaches
            for analyzing or integrating these data? Do you want to help improve
            the utility and accessibility of single-cell and spatial omics
            data?</span
        >
    </p>
    <p class="c4">
        <span class="c1"
            >Join colleagues from a wide range of fields and communities to
            explore new approaches to analyze and integrate </span
        ><span class="c15 c1"
            ><a class="c7" href="https://humantumoratlas.org/explore"
                >single-cell and spatial omics data from the Human Tumor Atlas
                Network (HTAN)</a
            ></span
        ><span class="c1"
            >. By participating in this event, you will have the opportunity to
            learn more about the </span
        ><span class="c1"
            >HTAN data and related resources, help improve the utility of HTAN
            data, and help build a broader and more diverse HTAN
            community.</span
        >
    </p>
    <p class="c4 c17"><span class="c5 c1"></span></p>
    <p class="c4"><span class="c5 c1">About HTAN</span></p>
    <p class="c4">
        <span class="">The goal of the </span
        ><span class="c15"
            ><a class="c7" href="https://humantumoratlas.org/"
                >Human Tumor Atlas Network</a
            ></span
        ><span class=""
            >&nbsp;(HTAN) is to generate molecularly defined, multi-dimensional
            spatial atlases of tumors as they progress from precancer to cancer,
            transition from locally invasive to metastatic disease, and respond
            or become resistant to therapy. A brief overview of HTAN can be seen </span
        ><span class="c21 c1"
            ><a class="c7" href="https://videocast.nih.gov/watch%3D45619"
                >here</a
            ></span
        ><span class="c1"
            >&nbsp;(starting at 02h:11m:56s). The HTAN includes a Data
            Coordinating Center, five precancer </span
        ><span class="c1">&nbsp;atlas generating centers</span
        ><span class="c1"
            >, and five advanced cancer atlas generating centers. You can read
            more about the cancer and scientific focus of each center </span
        ><span class="c15 c1"
            ><a class="c7" href="https://humantumoratlas.org/research-network"
                >here</a
            ></span
        ><span class="c1">. </span
        ><span class="c0"
            >The initial set of HTAN atlases provide several new insights into
            the biology of precancer, metastasis, and therapy resistance and
            suggest opportunities for translation of a number HTAN findings.
        </span>
    </p>
    <p class="c4">
        <span class="c1 c30">Importantly, </span
        ><span class="c1"
            >HTAN also operates as a community resource and extensive efforts
            have been made across HTAN to provide the data, protocols, and
            analytical methods underlying the HTAN atlases available to the
            community. You can explore the available HTAN data across multiple
            precancer and cancer types, including breast, colon, lung, skin,
            pancreas, and pediatric leukemia, </span
        ><span class="c15 c1"
            ><a class="c7" href="https://humantumoratlas.org/explore"
                >here</a
            ></span
        ><span class="c0">.</span>
    </p>
    <p class="c4 c17"><span class="c0"></span></p>
    <p class="c4">
        <span class="c5 c1">About the HTAN Data Jamborees</span>
    </p>
    <p class="c4">
        <span class="c1"
            >Data Jamborees provide amazing opportunities for scientists and
            coders of diverse expertise and experience levels to work together
            creatively to build unique solutions</span
        ><span class="c1"
            >&nbsp;that solve relevant problems in cancer research. &nbsp;The
            key to our Data Jamboree is the focus on collaboration. &nbsp;Over 4
            half-days, teams of four to eight people work on different
            scientific problems, emphasizing sharing expertise and insight
            through shared </span
        ><span class="c1">presentations</span
        ><span class="c0"
            >&nbsp;and live demonstrations throughout the event. &nbsp;The event
            will wrap up with a summary of the work accomplished during the
            event.</span
        >
    </p>
    <p class="c4">
        <span class="c1"
            >Participants may leverage the NCI CRDC resources. CRDC resources
            can provide an environment to work with HTAN data or combine HTAN
            data with </span
        ><span class="c1">publicly available datasets</span
        ><span class="c1"
            >, including those generated by Jamboree participants</span
        ><span class="c1"
            >. The HTAN atlas dataset contains clinical data matched to a wide
            variety of molecular data types, including bulk and single-cell
            omics and a variety of multiplexed spatial data types (i.e.,
            multiplexed immunofluorescence or spatial transcriptomics). Cloud
            credits will be available for use on the </span
        ><span class="c1">NCI CRDC Cloud</span
        ><span class="c1">&nbsp;Resource, </span
        ><span class="c21 c1"
            ><a class="c7" href="https://www.cancergenomicscloud.org/"
                >Seven Bridges Cancer Genomics Cloud</a
            ></span
        ><span class="c1"
            >, to continue project development and to nurture nascent
            collaborations.</span
        >
    </p>
    <p class="c4">
        <span class="c1"
            >The HTAN Data Jamborees aim to bring together investigators from
            outside of HTAN with an interest in utilizing HTAN data and
            associated tools to partner with HTAN investigators. </span
        ><span class="c1 c8 c23"
            >The Jamboree planning team will organize par</span
        ><span class="c1 c8 c23">ticipants </span
        ><span class="c1 c8 c23"
            >into small teams with diverse expertise and experience. Each team
            will be focused on addressing a specific question or challenge
            utilizing available HTAN data.</span
        ><span class="c0"
            >&nbsp;Teams will work together virtually over 4 half-days to
            develop creative solutions leveraging HTAN data and resources. The
            goals of the HTAN Data Jamborees include:</span
        >
    </p>
    <ol class="c12 lst-kix_list_5-0 start" start="1">
        <li class="c11 li-bullet-0">
            <span class="c0">Promote the </span
            ><span class="c15 c1 c24"
                ><a
                    class="c7"
                    href="https://www.nature.com/articles/sdata201618"
                    >FAIR</a
                ></span
            ><span class="c0">&nbsp;access and reuse of HTAN data</span>
        </li>
        <li class="c11 li-bullet-0">
            <span class="c0"
                >Promote collaborations to expand the size and diversity of the
                HTAN community</span
            >
        </li>
        <li class="c11 li-bullet-0">
            <span class="c0">Promote development of </span
            ><span class="c0">new methods</span
            ><span class="c0">&nbsp;and tools for HTAN data analysis</span>
        </li>
        <li class="c4 c27 li-bullet-0">
            <span class="c0"
                >Identify gaps and limitations of existing HTAN data and
                resources</span
            >
        </li>
    </ol>
    <p class="c4 c17"><span class="c0"></span></p>
    <p class="c4">
        <span class="c5 c1">HTAN Data Jamboree Projects</span>
    </p>
    <p class="c4">
        <span class="c1"
            >This HTAN Data Jamboree will focus on a variety of activities using
            the HTAN atlas datasets </span
        ><span class="c1"
            >with an emphasis on promoting the accessibility and reuse of HTAN
            data. &nbsp;</span
        ><span class="c1"
            >We are especially interested in participants that would like to
            launch new collaborative projects utilizing HTAN or HTAN-like
            datasets. &nbsp;Projects could include building analysis pipelines,
            visualization techniques, AI/ML algorithm development, employing
            statistical methods, or using existing computational, mathematical,
            or informatics tools to answer a new cancer-related question.
            Projects involving data interoperability, such as bringing public
            data to analyze with HTAN data are welcome. &nbsp;Projects to
            generate tutorial pipelines and educational tools are of interest.
            &nbsp;We are also interested in data storytelling, infographics, and
            other creative uses of the HTAN atlas data. </span
        ><span class="c1 c8"
            >Interested applicants with creative project ideas are encouraged to
            submit an optional project pitch with their application.</span
        >
    </p>
    <p class="c4 c17"><span class="c5 c1"></span></p>
    <p class="c4"><span class="c5 c1">Team Leads</span></p>
    <p class="c4">
        <span class="c1"
            >Whether or not you propose a project idea, you can indicate your
            interest in serving as a team </span
        ><span class="c1">lead</span
        ><span class="c0"
            >. Each team lead will coordinate a group of 4-8 people in defining
            the project and producing a clear vision for developing a solution.
            To accomplish this goal, the team lead must define and delegate
            tasks, incorporate team members&rsquo; ideas to accomplish the goal
            and ensure the team&rsquo;s success. Team leads are not responsible
            for forming teams. The Data Jamboree planning team will create teams
            of 4-8 individuals with various backgrounds and relevant expertise
            to work on each project. Each team will have access to mentors with
            technical expertise and working knowledge of the HTAN Atlas
            datasets. All participants in the event, including team leads, will
            need to be present for all 4 half-days.</span
        >
    </p>
    <p class="c4 c17"><span class="c5 c1"></span></p>
    <p class="c4"><span class="c5 c1">Who Should Apply</span></p>
    <p class="c4">
        <span class="c1"
            >Independent investigators, postdocs, and late-stage graduate
            students interested in utilizing HTAN single-cell and spatial omics
            data as part of their research. Investigators with cancer biology,
            cancer clinical research, systems biology, computational biology,
            and other relevant fields are all encouraged to apply. </span
        ><span class="c1 c8 c31"
            >Computational expertise is not required to participate</span
        ><span class="c1 c31">.</span
        ><span class="c0"
            >&nbsp;Participants will be organized into teams with diverse
            expertise by the organizing committee.</span
        >
    </p>
    <p class="c4 c17"><span class="c5 c1"></span></p>
    <p class="c4"><span class="c5 c1">Data Jamboree Schedule</span></p>
    <p class="c4">
        <span class="c1"
            >The HTAN Data Jamboree will run from Monday, December 4 through
            Thursday, December 7. Monday through Wednesday, we will kick off
            each day as a group at 11am Eastern Standard Time (EST) and wrap up
            each day around 5pm EST. Thursday will focus on final demonstrations
            from each group, and we will start at 1pm EST and wrap up by 4pm
            EST.</span
        >
    </p>
    <p class="c4 c17"><span class="c5 c32"></span></p>
    <p class="c4"><span class="c1 c8">How To Apply</span></p>
    <p class="c4">
        <span class="c1">Please&nbsp;submit your application by </span
        ><span class="c33 c1">October 24, 2023 </span
        ><span class="c1">using the </span
        ><span class="c21 c1"
            ><a class="c7" href="https://forms.gle/V2XW6bm2KJJbX6CF6"
                >Application Form</a
            ></span
        ><span class="c1 c26">.</span>
    </p>
    <p class="c4 c17"><span class="c5 c1"></span></p>
    <p class="c4">
        <span class="c1 c8">Questions</span><span class="c1 c8">?</span>
    </p>
    <p class="c4">
        <span class="c1"
            >For questions about the HTAN Data Jamborees in general or about the
            application process, please email Sean Hanlon (</span
        ><span class="c21 c1"
            ><a class="c7" href="mailto:sean.hanlon@nih.gov"
                >sean.hanlon@nih.gov</a
            ></span
        ><span class="c1">)</span>
    </p>
</body>
